backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+		}
+		if srv.locks > maxServerLocks || now.After(srv.expiry) {
+			continue // Overlocked or expired server.
+		}
+		srv.locks++
+		fs.Debugf(f, "Lock server #%d %q with %d locks", i, srv.url, srv.locks)
+		return srv.url, nil
+	}
+
+	// Server not found - ask Mailru dispatcher.
+	opts := rest.Opts{
+		Method:  "GET",
+		RootURL: api.DispatchServerURL,
+		Path:    sl.path,
+	}
+	var (
+		res *http.Response
+		url string
+		err error
+	)
+	err = f.pacer.Call(func() (bool, error) {
+		res, err = f.srv.Call(&opts)
+		if err != nil {
+			return fserrors.ShouldRetry(err), err
+		}
+		url, err = readBodyWord(res)
+		return fserrors.ShouldRetry(err), err
+	})
+	if err != nil || url == "" {
+		closeBody(res)
+		return "", errors.Wrap(err, "Failed to request file server")
+	}
+
+	// Attach to a server proposed by dispatcher.
+	for i := 0; i < num; i++ {
+		srv := &sl.list[i]
+		if srv.url != url {
+			continue // Not a proposed server.
+		}
+		srv.locks++
+		srv.expiry = now.Add(sl.expirySec * time.Second)
+		if fs.Config.LogLevel >= fs.LogLevelInfo {
+			ctime, _ := srv.expiry.MarshalJSON()
+			fs.Debugf(f, "Attach server #%d %q with %d locks (new expiry %s)", i, url, srv.locks, ctime)
+		}
+		return url, nil
+	}
+
+	// Find a free or unused server slot.
+	var server *pendingServer
+	var i int
+	for i = 0; i < num; i++ {
+		srv := &sl.list[i]
+		if srv.url == "" || srv.locks <= 0 {
+			server = srv
+			break
+		}
+	}
+	if server == nil {
+		// No free slots - extend the array.
+		sl.list = append(sl.list, pendingServer{})
+		num = len(sl.list)
+		i = num - 1
+		server = &sl.list[i]
+	}
+
+	server.expiry = now.Add(sl.expirySec * time.Second)
+	if fs.Config.LogLevel >= fs.LogLevelDebug {
+		ctime, _ := server.expiry.MarshalJSON()
+		fs.Debugf(f, "Switch server #%d %q -> %q with locks %d -> 1 (expires %s)", i, server.url, url, server.locks, ctime)
+	}
+	server.url = url
+	server.locks = 1
+	return url, nil
+}
+
+func (sl *serverList) free(url string) {
+	if url == "" {
+		return
+	}
+	sl.mu.Lock()
+	defer sl.mu.Unlock()
+
+	for i := 0; i < len(sl.list); i++ {
+		srv := &sl.list[i]
+		if srv.url != url {
+			continue // Not our server
+		}
+		if srv.locks > 0 {
+			fs.Debugf(sl.fs, "Unlock server #%d %q with %d locks", i, srv.url, srv.locks)
+			srv.locks--
+		} else {
+			// Getting here indicates possible race
+			fs.Infof(sl.fs, "Free server #%d %q", i, srv.url)
+			srv.url = ""
+			srv.locks = 0
+		}
+	}
+}
+
+// Name of the remote (as passed into NewFs)
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root of the remote (as passed into NewFs)
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// String converts this Fs to a string
+func (f *Fs) String() string {
+	return fmt.Sprintf("[%s]", f.root)
+}
+
+// Precision return the precision of this Fs
+func (f *Fs) Precision() time.Duration {
+	return time.Second
+}
+
+// Hashes returns the supported hash sets
+func (f *Fs) Hashes() hash.Set {
```

Comment:
I think adding your patch to `fs/hash` and making this return an mrhash would be a good idea - it will mean rclone can check hashes on upload/download and using `rclone check`.